File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        // go as far left as possible
        // check the node off
        // check the parent off
        // check the right side off
        
        return helper(root, k);
        
    }
    
    int helper(TreeNode* root, int& k) {
        if(root == nullptr) {
            return -1;
        }
        else if(k == 0) {
            return root->val;
        }
        else{
            int a = helper(root->left, k);
            if(a != -1) {return a;}
            k--; // serves to check current node
            if(k == 0) {return root->val;}
            int b = helper(root->right, k);
            if(b != -1) {return b;}
            return -1;
        }
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
# def init(self, val=0, left=None, right=None):
# self.val = val
# self.left = left
# self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:

        self.currK = 0
        self.finalValue = 0

        def recursiveTreeTraversal(currNode):

            if currNode:
                recursiveTreeTraversal(currNode.left)
            else:
                return
            self.currK += 1
            print(self.currK)
            
            if self.currK == k:
                print("Kth smallest value:", currNode.val)
                
                self.finalValue = currNode.val
                self.currK += 1
                return

            if currNode.right:
                recursiveTreeTraversal(currNode.right)

        recursiveTreeTraversal(root)

        return self.finalValue
